### PR TITLE
_is_peft_model update to recognise peft submodules, allowing training quantised models with peft submodules

### LINKED ISSUE #30878: Have `_is_peft_model` check if there's any peft submodule/Allow quantised training
### Feature request

I have multi-modal models with multiple different peft models on different submodules due to requiring different LoRA configurations. The following is an example:

```python
model = AutoModelForVision2Seq.from_pretrained(
  args.pretrained_ckpt,
  torch_dtype=compute_dtype,
  quantization_config=BitsAndBytesConfig(
      load_in_4bit=bits == 4,
      load_in_8bit=bits == 8,
      llm_int8_threshold=6.0,
      int8_quant_skip_modules=int8_quant_skip_modules,
      llm_int8_has_fp16_weight=False,
      bnb_4bit_compute_dtype=compute_dtype,
      bnb_4bit_use_double_quant=True,
      bnb_4bit_quant_type='nf4'  # {'fp4', 'nf4'}
  ) if bits < 16 else None,
  attn_implementation=args.attn_implementation,
)

if (args.use_lora and not resume_from_checkpoint and not ft_checkpoint_dir):
  target_modules = get_target_modules(model.model.text_model, args, bits)
  peft_config = LoraConfig(
      target_modules=target_modules,
      inference_mode=args.inference_mode,
      r=args.lora_r,
      lora_alpha=args.lora_alpha,
      lora_dropout=args.lora_dropout,
      use_dora=args.use_dora
  )
  model.model.text_model = get_peft_model(model.model.text_model, peft_config)

  if args.vit_train:
      target_modules = get_target_modules(model.model.vision_model, args, args.vit_bits, vit=True)
      peft_config = LoraConfig(
          target_modules=target_modules,
          inference_mode=args.inference_mode,
          r=args.vit_lora_r,
          lora_alpha=args.vit_lora_alpha,
          lora_dropout=args.lora_dropout,
          use_dora=args.use_dora_vit
      )
      model.model.vision_model = get_peft_model(model.model.vision_model, peft_config)

  if args.lora_abstractor:
      target_modules = get_target_modules(model.model.connector, args, args.bits)
      peft_config = LoraConfig(
          target_modules=target_modules,
          inference_mode=args.inference_mode,
          r=args.lora_r,
          lora_alpha=args.lora_alpha,
          lora_dropout=args.lora_dropout,
          use_dora=args.use_dora
      )
      model.model.connector = get_peft_model(model.model.connector, peft_config)

```
This works fine until you quantise the model as the huggingface trainer requires that the model pass the `_is_peft_model` check when it's quantised or it assumes the entire model is quantised and therefore untrainable. A really simple fix would just to be to change it from a check with `_is_peft_model` to just seeing if there are any parameters that are unquantised AND of a trainable datatype (>=16bit) but if it really should stick to being a peft model check it could just check if there are any submodules which are peft models. I think in general, allowing functionality to generalise irrespective of what class or model type the top level model is will just generally be helpful for researchers in the multi-modal space such as myself.

I understand that there is PeftMixedModel but it seems largely unsupported and unwieldy compared to just doing it like this where I have complete control.

### Motivation

Let me train quantised models that aren't PeftModels at their top level/aren't wrapped as a PeftModel.

### Your contribution

Here's a version of `_is_peft_model` that uses param requiring gradient and not being 4 or 8 bit
```python
def _is_peft_model(model):
    if is_peft_available():
        classes_to_check = (PeftModel,) if is_peft_available() else ()
        # Here we also check if the model is an instance of `PeftMixedModel` introduced in peft>=0.7.0: https://github.com/huggingface/transformers/pull/28321
        if version.parse(importlib.metadata.version("peft")) >= version.parse("0.7.0"):
            from peft import PeftMixedModel

            classes_to_check = (*classes_to_check, PeftMixedModel)
        is_peft = isinstance(model, classes_to_check)
        if is_peft:
            return True
        else:
            for _, param in model.named_parameters():
                if param.requires_grad and not "4bit" not in param.__class__.__name__ and "8bit" not in param.__class__.__name__:
                    return True
    return False
```

Here is probably a more acceptable version that specifically checks for any PeftModel submodules:
```python
def _is_peft_model(model):
    if is_peft_available():
        classes_to_check = (PeftModel,) if is_peft_available() else ()
        # Here we also check if the model is an instance of `PeftMixedModel` introduced in peft>=0.7.0: https://github.com/huggingface/transformers/pull/28321
        if version.parse(importlib.metadata.version("peft")) >= version.parse("0.7.0"):
            from peft import PeftMixedModel

            classes_to_check = (*classes_to_check, PeftMixedModel)
        is_peft = isinstance(model, classes_to_check)
        if is_peft:
            return True
        else:
            for submodule in model.modules():
                if isinstance(submodule, classes_to_check):
                    return True
    return False
```

### PR DESCRIPTION
**Pull request by ambroser53**
# What does this PR do?

Don't necessarily have peft models as the top-level wrapper for models, especially when working with custom built multi-modal models. For example:

```python
model = AutoModelForVision2Seq.from_pretrained(
  args.pretrained_ckpt,
  torch_dtype=compute_dtype,
  quantization_config=BitsAndBytesConfig(
      load_in_4bit=bits == 4,
      load_in_8bit=bits == 8,
      llm_int8_threshold=6.0,
      int8_quant_skip_modules=int8_quant_skip_modules,
      llm_int8_has_fp16_weight=False,
      bnb_4bit_compute_dtype=compute_dtype,
      bnb_4bit_use_double_quant=True,
      bnb_4bit_quant_type='nf4'  # {'fp4', 'nf4'}
  ) if bits < 16 else None,
  attn_implementation=args.attn_implementation,
)

if (args.use_lora and not resume_from_checkpoint and not ft_checkpoint_dir):
  target_modules = get_target_modules(model.model.text_model, args, bits)
  peft_config = LoraConfig(
      target_modules=target_modules,
      inference_mode=args.inference_mode,
      r=args.lora_r,
      lora_alpha=args.lora_alpha,
      lora_dropout=args.lora_dropout,
      use_dora=args.use_dora
  )
  model.model.text_model = get_peft_model(model.model.text_model, peft_config)

  if args.vit_train:
      target_modules = get_target_modules(model.model.vision_model, args, args.vit_bits, vit=True)
      peft_config = LoraConfig(
          target_modules=target_modules,
          inference_mode=args.inference_mode,
          r=args.vit_lora_r,
          lora_alpha=args.vit_lora_alpha,
          lora_dropout=args.lora_dropout,
          use_dora=args.use_dora_vit
      )
      model.model.vision_model = get_peft_model(model.model.vision_model, peft_config)

  if args.lora_abstractor:
      target_modules = get_target_modules(model.model.connector, args, args.bits)
      peft_config = LoraConfig(
          target_modules=target_modules,
          inference_mode=args.inference_mode,
          r=args.lora_r,
          lora_alpha=args.lora_alpha,
          lora_dropout=args.lora_dropout,
          use_dora=args.use_dora
      )
      model.model.connector = get_peft_model(model.model.connector, peft_config)
```

This allows the hf trainer to recognise such models as still being peft models and thereby allow quantised training (QLoRA).

Fixes #30878

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. -> https://github.com/huggingface/transformers/issues/30878
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@younesbelkada

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -247,7 +247,13 @@ def _is_peft_model(model):
             from peft import PeftMixedModel
 
             classes_to_check = (*classes_to_check, PeftMixedModel)
-        return isinstance(model, classes_to_check)
+        is_peft = isinstance(model, classes_to_check)
+        if is_peft:
+            return True
```
Comment (younesbelkada) [inline src/transformers/trainer.py:252]:
```suggestion
        if isinstance(model, classes_to_check):
            return True
```

**Review by younesbelkada — COMMENTED** [review]
Thanks for adding the submodules support for PEFT + Trainer ! Left one suggestion - what do you think?

File: src/transformers/trainer.py (hunk 1 above)
Comment (ambroser53) [reply src/transformers/trainer.py]: Good catch. Made the change.

Comment (younesbelkada) [reply src/transformers/trainer.py]: thanks @ambroser53 !

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding! Could you add a test with a dummy model that has peft submodules that behaves correctly with this change but fails on main?

**Review by younesbelkada — COMMENTED** [review]
Thanks @ambroser53 ! 
For the styling checks, can you try to run `pip install -U ".[quality]"` and re-run `make fixup` ?

**Comment by ambroser53** [comment]
@younesbelkada I've just done as you asked and it ran successfully but I have no working tree changes to commit.

**Comment by younesbelkada** [comment]
Thanks ! Hmm I think something is off with our CI currently, let's wait for https://github.com/huggingface/transformers/pull/30932 being merged first

File: tests/trainer/test_trainer.py [hunk 2]
```diff
@@ -989,6 +990,40 @@ def test_bnb_compile(self):
...
                 _ = Trainer(tiny_model, args, train_dataset=train_dataset)  # noqa
 
+
+    @require_peft
+    @require_bitsandbytes
+    def test_peft_submodule(self):
+        from peft import LoraConfig, get_peft_model
+
+        # Simply tests if initializing a Trainer with a PEFT on a submodule will pass _is_peft_model check.
+        # Throws error when quantised but not recognised as a peft model.
```
Comment (amyeroberts) [inline tests/trainer/test_trainer.py:1000]: Could you clarify this comment a bit? I can't parse `"Throws error when quantised but not recognised as a peft model."`. Specifically, the "but" formulation makes it unclear the intended behaviour and causation (is it throwing an error because it's quantized? Should it throw an error when quantized? Should it recognize it's a peft model?)

File: tests/trainer/test_trainer.py [hunk 3]
```diff
@@ -989,6 +990,40 @@ def test_bnb_compile(self):
...
+        train_dataset = RepeatDataset(x)
+
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            args = TrainingArguments(
+                tmp_dir,
+                learning_rate=1e-9,
+                logging_steps=5,
+            )
+            with self.assertRaises(ValueError):
+                _ = Trainer(tiny_model, args, train_dataset=train_dataset)  # noqa
```
Comment (amyeroberts) [inline tests/trainer/test_trainer.py:1024]: Wait - I'm confused 😅 

My understanding of this PR was that it's meant to allow training on quantized models which have a PEFT submodule. 

This tests doesn't obviously quantize (I might be missing something here). Is it meant to fail if the model isn't quantized?

**Review by amyeroberts — COMMENTED** [review]
Thanks for iterating on this. 

I'm a bit confused about the intended behaviour here from the tests

File: tests/trainer/test_trainer.py (hunk 3 above)
Comment (ambroser53) [reply tests/trainer/test_trainer.py:1024]: It does quantise the model. See `load_in_4bit=True` on line 1002.

File: tests/trainer/test_trainer.py (hunk 2 above)
Comment (ambroser53) [reply tests/trainer/test_trainer.py:1000]: "but" is meant to be "as"

File: tests/trainer/test_trainer.py (hunk 3 above)
Comment (amyeroberts) [reply tests/trainer/test_trainer.py:1024]: Ah, sorry, I missed that. In this case, why is this error being thrown? Shouldn't this model now be trainable (it's quantized and has a peft submodule).

**Comment by ambroser53** [comment]
Is there any update on this? I'm still running the local version for training my model that uses this. Still don't understand why it couldn't be merged in the first place.

**Comment by amyeroberts** [comment]
@ambroser53 It's still open as there's two outstanding comments/questions on the tests which were marked as resolved, but were not. In the first case, the comment needs to be updated, in the second, could you explain the intended behaviour?

**Comment by ambroser53** [comment]
@amyeroberts I see where the confusion was here. I was under the assumption the code was fine and we were waiting on a fix for workflow checks (i.e. styling). I have changed the comments in the test to hopefully be more clear and expanded the it to run `trainer.train()` instead of just initialising the Trainer object and included a simple custom subclass with `save_pretrained` to allow it to work in that case to show the full use-case with training. 

Intended Behaviour: model is trainable when quantised with a peft sub-module.

Workflow tests are still failing so if there is actually anything I need to do there let me know.

**Comment by amyeroberts** [comment]
@ambroser53 Great. For some of the failing tests re passing `trust_remote_code` for datasets, there's been upstream fixes. Could you rebase on main to include these? I think this should also fix a lot of the quality checks

**Comment by amyeroberts** [comment]
@ambroser53 Apologies for the continued delay. Could you try rebasing again? This should resolve the timeout errors we're having on the CI

File: tests/trainer/test_trainer.py [hunk 4]
```diff
@@ -996,6 +997,49 @@ def test_bnb_compile(self):
...
                 _ = Trainer(tiny_model, args, train_dataset=train_dataset)  # noqa
 
+    @require_peft
+    @require_bitsandbytes
+    def test_peft_submodule(self):
+        from peft import LoraConfig, PeftModel, get_peft_model
+
+        # Due to the way the Trainer is implemented we must be able to save the model with 'save_pretrained'
+        # Therefore to use peft submodules you must specify your own 'save_pretrained' method
+        # This example subclass allows for the saving of any and all submodules that are of type PeftModel
```
Comment (amyeroberts) [inline tests/trainer/test_trainer.py:1007]: This isn't right as it means we're no longer testing the transformers functionality. Instead, this PR should update `save_pretrained` to enable saving of models with PEFT submodules 

Not sure who's best to tag here @BenjaminBossan @SunMarc

**Review by amyeroberts — COMMENTED** [review]
Thanks for iterating! 

As mentioned in my PR - needing to override `save_pretrained` in the test is an indication we should update the standard `save_pretrained` to account for models with peft submodules

File: tests/trainer/test_trainer.py (hunk 4 above)
Comment (BenjaminBossan) [reply tests/trainer/test_trainer.py:1007]: I would suggest splitting that issue off into a separate PR. IIUC, this PR is originally about avoiding a false positive `ValueError` that the model is presumably not trainable. I think we should focus on that here.

Now this test touches on another issue, namely that `save_pretrained` of a PEFT model only saves the trainable adapter part, but of the PEFT model itself is just a submodule, `save_pretrained` acts like it would on a normal transformers model and saves the whole checkpoint (unless I misunderstand the intent).

IMO, first of all, this should not be handled here in the test. Instead, I'd just add a comment that says something along the lines of: "Note that save_pretrained saves the whole model here, not just the PEFT adater". The test should still pass, right?

Regarding the question of what the right thing to do is if the model has PEFT submodule(s): IMO this is not quite clear and changing `save_pretrained` for all transformers users to only save the PEFT submodules could be breaking existing code. Probably 90% of users in this situation would only want to save the PEFT adapters, but they would already require special handling to load these submodules correctly (right?) so maybe it's fine that they also need special handling to save them? IMHO, If we want to make it possible to only save the PEFT adapaters of the submodules, it should be configurable, with the default being the status quo.
